File: Simplified/NYPLAccount.m
//#import "NYPLKeychain.h"
//
//#import "NYPLAccount.h"
//
//#import "SimplyE-Swift.h"
//
//NSString * authorizationIdentifierKey = @"NYPLAccountAuthorization";
//NSString * barcodeKey = @"NYPLAccountBarcode";
//NSString * PINKey = @"NYPLAccountPIN";
//NSString * adobeTokenKey = @"NYPLAccountAdobeTokenKey";
//NSString * licensorKey = @"NYPLAccountLicensorKey";
//NSString * patronKey = @"NYPLAccountPatronKey";
//NSString * authTokenKey = @"NYPLAccountAuthTokenKey";
//NSString * adobeVendorKey = @"NYPLAccountAdobeVendorKey";
//NSString * providerKey = @"NYPLAccountProviderKey";
//NSString * userIDKey = @"NYPLAccountUserIDKey";
//NSString * deviceIDKey = @"NYPLAccountDeviceIDKey";
//
//
//@implementation NYPLAccount
//
//+ (instancetype)sharedAccount
//{
//  NSString *library = [AccountsManager shared].currentAccount.uuid;
//  return [self sharedAccount:library];
//}
//
//+ (instancetype)sharedAccount:(NSString *)libraryUUID
//{
//  static NYPLAccount *sharedAccount = nil;
//  
//  if (sharedAccount == nil) {
//    sharedAccount = [[self alloc] init];
//    if(!sharedAccount) {
//      NYPLLOG(@"Failed to create shared account.");
//    }
//  }
//  
//  if (![libraryUUID isEqualToString:[AccountsManager NYPLAccountUUIDs][0]])
//  {
//    barcodeKey = [NSString stringWithFormat:@"%@_%@",@"NYPLAccountBarcode", libraryUUID];
//    authorizationIdentifierKey = [NSString stringWithFormat:@"%@_%@",@"NYPLAccountAuthorization", libraryUUID];
//    PINKey = [NSString stringWithFormat:@"%@_%@",@"NYPLAccountPIN", libraryUUID];
//    adobeTokenKey = [NSString stringWithFormat:@"%@_%@",@"NYPLAccountAdobeTokenKey", libraryUUID];
//    patronKey = [NSString stringWithFormat:@"%@_%@",@"NYPLAccountPatronKey", libraryUUID];
//    authTokenKey = [NSString stringWithFormat:@"%@_%@",@"NYPLAccountAuthTokenKey", libraryUUID];
//    adobeVendorKey = [NSString stringWithFormat:@"%@_%@",@"NYPLAccountAdobeVendorKey", libraryUUID];
//    providerKey = [NSString stringWithFormat:@"%@_%@",@"NYPLAccountProviderKey", libraryUUID];
//    userIDKey = [NSString stringWithFormat:@"%@_%@",@"NYPLAccountUserIDKey", libraryUUID];
//    deviceIDKey = [NSString stringWithFormat:@"%@_%@",@"NYPLAccountDeviceIDKey", libraryUUID];
//    licensorKey = [NSString stringWithFormat:@"%@_%@",@"NYPLAccountLicensorKey", libraryUUID];
//  }
//  else
//  {
//    barcodeKey = @"NYPLAccountBarcode";
//    authorizationIdentifierKey = @"NYPLAccountAuthorization";
//    PINKey = @"NYPLAccountPIN";
//    adobeTokenKey = @"NYPLAccountAdobeTokenKey";
//    patronKey = @"NYPLAccountPatronKey";
//    authTokenKey = @"NYPLAccountAuthTokenKey";
//    adobeVendorKey = @"NYPLAccountAdobeVendorKey";
//    providerKey = @"NYPLAccountProviderKey";
//    userIDKey = @"NYPLAccountUserIDKey";
//    deviceIDKey = @"NYPLAccountDeviceIDKey";
//    licensorKey = @"NYPLAccountLicensorKey";
//  }
//
//  return sharedAccount;
//}
//
//- (BOOL)hasCredentials
//{
//  return self.hasAuthToken || self.hasBarcodeAndPIN;
//}
//
//- (BOOL)hasBarcodeAndPIN
//{
//  return self.barcode != nil && self.PIN != nil;
//}
//
//- (BOOL)hasAuthToken
//{
//  return self.authToken != nil;
//}
//
//- (BOOL)hasAdobeToken
//{
//  return self.adobeToken != nil;
//}
//
//- (BOOL)hasLicensor
//{
//  return self.licensor != nil;
//}
//
//- (NSString *)authorizationIdentifier
//{
//  return [[NYPLKeychain sharedKeychain] objectForKey:authorizationIdentifierKey];
//}
//
//- (NSString *)barcode
//{
//  return [[NYPLKeychain sharedKeychain] objectForKey:barcodeKey];
//}
//
//- (NSString *)PIN
//{
//  return [[NYPLKeychain sharedKeychain] objectForKey:PINKey];
//}
//
//- (NSString *)adobeVendor
//{
//  return [[NYPLKeychain sharedKeychain] objectForKey:adobeVendorKey];
//}
//
//- (NSString *)adobeToken
//{
//  return [[NYPLKeychain sharedKeychain] objectForKey:adobeTokenKey];
//}
//
//- (NSDictionary *)licensor
//{
//  return [[NYPLKeychain sharedKeychain] objectForKey:licensorKey];
//}
//
//- (NSDictionary *)patron
//{
//  return [[NYPLKeychain sharedKeychain] objectForKey:patronKey];
//}
//
//- (NSString *)patronFullName
//{
//  return [NSString stringWithFormat:@"%@ %@ %@", self.patron[@"name"][@"first"], self.patron[@"name"][@"middle"],self.patron[@"name"][@"last"]];
//}
//
//- (NSString *)authToken
//{
//  return [[NYPLKeychain sharedKeychain] objectForKey:authTokenKey];
//}
//
//- (NSString *)provider
//{
//  return [[NYPLKeychain sharedKeychain] objectForKey:providerKey];
//}
//
//- (NSString *)userID
//{
//  return [[NYPLKeychain sharedKeychain] objectForKey:userIDKey];
//}
//
//- (NSString *)deviceID
//{
//  return [[NYPLKeychain sharedKeychain] objectForKey:deviceIDKey];
//}
//
//- (void)setBarcode:(NSString *const)barcode PIN:(NSString *)PIN
//{
//  if(!(barcode && PIN)) {
//    @throw NSInvalidArgumentException;
//  }
//
//  [[NYPLKeychain sharedKeychain] setObject:barcode forKey:barcodeKey];
//  [[NYPLKeychain sharedKeychain] setObject:PIN forKey:PINKey];
//
//  // make sure to set the barcode related to the current account (aka library)
//  // not the one we just signed in to, because we could have signed in into
//  // library A, but still browsing the catalog of library B.
//  [NYPLErrorLogger setUserID:[[NYPLAccount sharedAccount] barcode]];
//  
//  [[NSNotificationCenter defaultCenter]
//   postNotificationName:NYPLAccountDidChangeNotification
//   object:self];
//}
//
//- (void)setAdobeToken:(NSString *)adobeToken patron:(NSDictionary *)patron
//{
//  if(!(adobeToken && patron)) {
//    @throw NSInvalidArgumentException;
//  }
//  
//  [[NYPLKeychain sharedKeychain] setObject:adobeToken forKey:adobeTokenKey];
//  [[NYPLKeychain sharedKeychain] setObject:patron forKey:patronKey];
//  
//  [[NSNotificationCenter defaultCenter]
//   postNotificationName:NYPLAccountDidChangeNotification
//   object:self];
//
//}
//- (void)setAdobeToken:(NSString *)adobeToken
//{
//  if(!(adobeToken)) {
//    @throw NSInvalidArgumentException;
//  }
//
//  [[NYPLKeychain sharedKeychain] setObject:adobeToken forKey:adobeTokenKey];
//  
//  [[NSNotificationCenter defaultCenter]
//   postNotificationName:NYPLAccountDidChangeNotification
//   object:self];
//
//}
//- (void)setLicensor:(NSDictionary *)licensor
//{
//  if(!(licensor)) {
//    @throw NSInvalidArgumentException;
//  }
//  
//  [[NYPLKeychain sharedKeychain] setObject:licensor forKey:licensorKey];
//}
//- (void)setAuthorizationIdentifier:(NSString *)identifier
//{
//  if(!(identifier)) {
//    @throw NSInvalidArgumentException;
//  }
//  
//  [[NYPLKeychain sharedKeychain] setObject:identifier forKey:authorizationIdentifierKey];
//}
//- (void)setPatron:(NSDictionary *)patron
//{
//  if(!(patron)) {
//    @throw NSInvalidArgumentException;
//  }
//
//  [[NYPLKeychain sharedKeychain] setObject:patron forKey:patronKey];
//  
//  [[NSNotificationCenter defaultCenter]
//   postNotificationName:NYPLAccountDidChangeNotification
//   object:self];
//
//}
//- (void)setAuthToken:(NSString *)authToken
//{
//  if(!(authToken)) {
//    @throw NSInvalidArgumentException;
//  }
//
//  [[NYPLKeychain sharedKeychain] setObject:authToken forKey:authTokenKey];
//  
//  [[NSNotificationCenter defaultCenter]
//   postNotificationName:NYPLAccountDidChangeNotification
//   object:self];
//}
//
//- (void)setAdobeVendor:(NSString *)adobeVendor
//{
//  if(!(adobeVendor)) {
//    @throw NSInvalidArgumentException;
//  }
//  
//  [[NYPLKeychain sharedKeychain] setObject:adobeVendor forKey:adobeVendorKey];
//  
//  [[NSNotificationCenter defaultCenter]
//   postNotificationName:NYPLAccountDidChangeNotification
//   object:self];
//}
//
//- (void)setProvider:(NSString *)provider
//{
//  if(!(provider)) {
//    @throw NSInvalidArgumentException;
//  }
//  
//  [[NYPLKeychain sharedKeychain] setObject:provider forKey:providerKey];
//  
//  [[NSNotificationCenter defaultCenter]
//   postNotificationName:NYPLAccountDidChangeNotification
//   object:self];
//}
//
//- (void)setUserID:(NSString *)userID
//{
//  if(!(userID)) {
//    @throw NSInvalidArgumentException;
//  }
//
//  [[NYPLKeychain sharedKeychain] setObject:userID forKey:userIDKey];
//  
//  [[NSNotificationCenter defaultCenter]
//   postNotificationName:NYPLAccountDidChangeNotification
//   object:self];
//}
//
//- (void)setDeviceID:(NSString *)deviceID
//{
//  if(!(deviceID)) {
//    @throw NSInvalidArgumentException;
//  }
//  
//  [[NYPLKeychain sharedKeychain] setObject:deviceID forKey:deviceIDKey];
//  
//  [[NSNotificationCenter defaultCenter]
//   postNotificationName:NYPLAccountDidChangeNotification
//   object:self];
//}
//
//- (void)removeAll
//{
//  [[NYPLKeychain sharedKeychain] removeObjectForKey:adobeTokenKey];
//  [[NYPLKeychain sharedKeychain] removeObjectForKey:patronKey];
//  [[NYPLKeychain sharedKeychain] removeObjectForKey:authTokenKey];
//  [[NYPLKeychain sharedKeychain] removeObjectForKey:adobeVendorKey];
//  [[NYPLKeychain sharedKeychain] removeObjectForKey:providerKey];
//  [[NYPLKeychain sharedKeychain] removeObjectForKey:userIDKey];
//  [[NYPLKeychain sharedKeychain] removeObjectForKey:deviceIDKey];
//
//  [self removeBarcodeAndPIN];
//}
//
//- (void)removeBarcodeAndPIN
//{
//  [[NYPLKeychain sharedKeychain] removeObjectForKey:barcodeKey];
//  [[NYPLKeychain sharedKeychain] removeObjectForKey:authorizationIdentifierKey];
//  [[NYPLKeychain sharedKeychain] removeObjectForKey:PINKey];
//
//  [[NSNotificationCenter defaultCenter]
//   postNotificationName:NYPLAccountDidChangeNotification
//   object:self];
//}
//
//@end

File: Simplified/NYPLAccount.swift
import Foundation

@objcMembers class NYPLAccount : NSObject {
  static private let shared = NYPLAccount()
    
  private var authorizationIdentifierKey = "NYPLAccountAuthorization"
  private var barcodeKey = "NYPLAccountBarcode"
  private var PINKey = "NYPLAccountPIN"
  private var adobeTokenKey = "NYPLAccountAdobeTokenKey"
  private var licensorKey = "NYPLAccountLicensorKey"
  private var patronKey = "NYPLAccountPatronKey"
  private var authTokenKey = "NYPLAccountAuthTokenKey"
  private var adobeVendorKey = "NYPLAccountAdobeVendorKey"
  private var providerKey = "NYPLAccountProviderKey"
  private var userIDKey = "NYPLAccountUserIDKey"
  private var deviceIDKey = "NYPLAccountDeviceIDKey"

  @objc class func sharedAccount() -> NYPLAccount
  {
    return sharedAccount(libraryUUID: AccountsManager.shared.currentAccount?.uuid)
  }
    
  @objc(sharedAccount:)
  class func sharedAccount(libraryUUID: String?) -> NYPLAccount
  {
    if let uuid = libraryUUID,
        uuid != AccountsManager.NYPLAccountUUIDs[0]
    {
      shared.barcodeKey = "NYPLAccountBarcode_\(uuid)"
      shared.authorizationIdentifierKey = "NYPLAccountAuthorization_\(uuid)"
      shared.PINKey = "NYPLAccountPIN_\(uuid)"
      shared.adobeTokenKey = "NYPLAccountAdobeTokenKey_\(uuid)"
      shared.patronKey = "NYPLAccountPatronKey_\(uuid)"
      shared.authTokenKey = "NYPLAccountAuthTokenKey_\(uuid)"
      shared.adobeVendorKey = "NYPLAccountAdobeVendorKey_\(uuid)"
      shared.providerKey = "NYPLAccountProviderKey_\(uuid)"
      shared.userIDKey = "NYPLAccountUserIDKey_\(uuid)"
      shared.deviceIDKey = "NYPLAccountDeviceIDKey_\(uuid)"
      shared.licensorKey = "NYPLAccountLicensorKey_\(uuid)"
    } else {
      shared.barcodeKey = "NYPLAccountBarcode"
      shared.authorizationIdentifierKey = "NYPLAccountAuthorization"
      shared.PINKey = "NYPLAccountPIN"
      shared.adobeTokenKey = "NYPLAccountAdobeTokenKey"
      shared.patronKey = "NYPLAccountPatronKey"
      shared.authTokenKey = "NYPLAccountAuthTokenKey"
      shared.adobeVendorKey = "NYPLAccountAdobeVendorKey"
      shared.providerKey = "NYPLAccountProviderKey"
      shared.userIDKey = "NYPLAccountUserIDKey"
      shared.deviceIDKey = "NYPLAccountDeviceIDKey"
      shared.licensorKey = "NYPLAccountLicensorKey"
    }
    
    return shared
  }
   
  // MARK: - Check
    
  func hasBarcodeAndPIN() -> Bool {
    return barcode != nil && PIN != nil
  }
  
  func hasAuthToken() -> Bool {
    return authToken != nil
  }
  
  func hasAdobeToken() -> Bool {
    return adobeToken != nil
  }
  
  func hasLicensor() -> Bool {
    return licensor != nil
  }
  
  func hasCredentials() -> Bool {
    return hasAuthToken() || hasBarcodeAndPIN()
  }
    
  // MARK: - GET
    
  var barcode: String? {
    return NYPLKeychain.shared()?.object(forKey: barcodeKey) as? String ?? nil
  }
    
  var authorizationIdentifier: String? {
    return NYPLKeychain.shared()?.object(forKey: authorizationIdentifierKey) as? String ?? nil
  }
  
  var PIN: String? {
    return NYPLKeychain.shared()?.object(forKey: PINKey) as? String ?? nil
  }
  
  var deviceID: String? {
    return NYPLKeychain.shared()?.object(forKey: deviceIDKey) as? String ?? nil
  }
  
  var userID: String? {
    return NYPLKeychain.shared()?.object(forKey: userIDKey) as? String ?? nil
  }
    
  var adobeVendor: String? {
    return NYPLKeychain.shared()?.object(forKey: adobeVendorKey) as? String ?? nil
  }
    
  var provider: String? {
    return NYPLKeychain.shared()?.object(forKey: providerKey) as? String ?? nil
  }
    
  var patron: [String:Any]? {
    return NYPLKeychain.shared()?.object(forKey: patronKey) as? [String:Any] ?? nil
  }
    
  var patronFullName: String? {
    if let patron = patron,
      let name = patron["name"] as? [String:String],
      let first = name["first"],
      let middle = name["middle"],
      let last = name["last"]
    {
      return "\(first) \(middle) \(last)"
    }
    return nil
  }
    
  var authToken: String? {
    return NYPLKeychain.shared()?.object(forKey: authTokenKey) as? String ?? nil
  }
    
  var adobeToken: String? {
    return NYPLKeychain.shared()?.object(forKey: adobeTokenKey) as? String ?? nil
  }
    
  var licensor: [String:Any]? {
    return NYPLKeychain.shared()?.object(forKey: licensorKey) as? [String:Any] ?? nil
  }
    
  // MARK: - SET
    
  @objc(setBarcode:PIN:)
  func setBarcode(barcode: String, PIN: String) {
    guard let sharedKeychain = NYPLKeychain.shared() else {
      return
    }
    
    sharedKeychain.setObject(barcode, forKey: barcodeKey)
    sharedKeychain.setObject(PIN, forKey: PINKey)
    
    // make sure to set the barcode related to the current account (aka library)
    // not the one we just signed in to, because we could have signed in into
    // library A, but still browsing the catalog of library B.
    NYPLErrorLogger.setUserID(NYPLAccount.sharedAccount().barcode)
    
    NotificationCenter.default.post(
      name: Notification.Name.NYPLAccountDidChange,
      object: self
    )
  }
    
  @objc(setAdobeToken:patron:)
  func setAdobeToken(token: String, patron: [String : Any]) {
    guard let sharedKeychain = NYPLKeychain.shared() else {
      return
    }
    
    sharedKeychain.setObject(token, forKey: adobeTokenKey)
    sharedKeychain.setObject(patron, forKey: patronKey)
    
    NotificationCenter.default.post(
      name: Notification.Name.NYPLAccountDidChange,
      object: self
    )
  }
  
  @objc(setAdobeVender:)
  func setAdobeVender(vendor: String) {
    guard let sharedKeychain = NYPLKeychain.shared() else {
      return
    }
    
    sharedKeychain.setObject(vendor, forKey: adobeVendorKey)
    
    NotificationCenter.default.post(
      name: Notification.Name.NYPLAccountDidChange,
      object: self
    )
  }
  
  @objc(setAdobeToken:)
  func setAdobeToken(token: String) {
    guard let sharedKeychain = NYPLKeychain.shared() else {
      return
    }
    
    sharedKeychain.setObject(token, forKey: adobeTokenKey)
    
    NotificationCenter.default.post(
      name: Notification.Name.NYPLAccountDidChange,
      object: self
    )
  }
  
  @objc(setLicensor:)
  func setLicensor(licensor: [String : Any]) {
    guard let sharedKeychain = NYPLKeychain.shared() else {
      return
    }
    
    sharedKeychain.setObject(licensor, forKey: licensorKey)
  }
  
  @objc(setAuthorizationIdentifier:)
  func setAuthorizationIdentifier(identifier: String) {
    guard let sharedKeychain = NYPLKeychain.shared() else {
      return
    }
    
    sharedKeychain.setObject(identifier, forKey: authorizationIdentifierKey)
  }
  
  @objc(setPatron:)
  func setPatron(patron: [String : Any]) {
    guard let sharedKeychain = NYPLKeychain.shared() else {
      return
    }
    
    sharedKeychain.setObject(patron, forKey: patronKey)
    
    NotificationCenter.default.post(
      name: Notification.Name.NYPLAccountDidChange,
      object: self
    )
  }
  
  @objc(setAuthToken:)
  func setAuthToken(token: String) {
    guard let sharedKeychain = NYPLKeychain.shared() else {
      return
    }
    
    sharedKeychain.setObject(token, forKey: authTokenKey)
    
    NotificationCenter.default.post(
      name: Notification.Name.NYPLAccountDidChange,
      object: self
    )
  }
  
  @objc(setProvider:)
  func setProvider(provider: String) {
    guard let sharedKeychain = NYPLKeychain.shared() else {
      return
    }
    
    sharedKeychain.setObject(provider, forKey: providerKey)
    
    NotificationCenter.default.post(
      name: Notification.Name.NYPLAccountDidChange,
      object: self
    )
  }
  
  @objc(setUserID:)
  func setUserID(id: String) {
    guard let sharedKeychain = NYPLKeychain.shared() else {
      return
    }
    
    sharedKeychain.setObject(id, forKey: userIDKey)
    
    NotificationCenter.default.post(
      name: Notification.Name.NYPLAccountDidChange,
      object: self
    )
  }
  
  @objc(setDeviceID:)
  func setDeviceID(id: String) {
    guard let sharedKeychain = NYPLKeychain.shared() else {
      return
    }
    
    sharedKeychain.setObject(id, forKey: deviceIDKey)
    
    NotificationCenter.default.post(
      name: Notification.Name.NYPLAccountDidChange,
      object: self
    )
  }
    
  // MARK: - Remove
    
  func removeBarcodeAndPIN() {
    guard let sharedKeychain = NYPLKeychain.shared() else {
      return
    }
    
    sharedKeychain.removeObject(forKey: barcodeKey)
    sharedKeychain.removeObject(forKey: authorizationIdentifierKey)
    sharedKeychain.removeObject(forKey: PINKey)
    
    NotificationCenter.default.post(
      name: Notification.Name.NYPLAccountDidChange,
      object: self
    )
  }
  
  func removeAll() {
    guard let sharedKeychain = NYPLKeychain.shared() else {
      return
    }
    
    sharedKeychain.removeObject(forKey: adobeTokenKey)
    sharedKeychain.removeObject(forKey: patronKey)
    sharedKeychain.removeObject(forKey: authTokenKey)
    sharedKeychain.removeObject(forKey: adobeVendorKey)
    sharedKeychain.removeObject(forKey: providerKey)
    sharedKeychain.removeObject(forKey: userIDKey)
    sharedKeychain.removeObject(forKey: deviceIDKey)
    
    removeBarcodeAndPIN()
  }
}
